lib/libc/string/bcopy.c

```diff
@@ -65,6 +71,104 @@ typedef	int word;		/* "word" used for optimal copy speed */
 #define	wsize	sizeof(word)
 #define	wmask	(wsize - 1)
 
+#if __has_feature(capabilities)
+#define write_to_stderr(str) __sys_write(STDERR_FILENO, (str), strlen(str))
+extern int __sysctlbyname(const char *name, size_t namelen, void *oldp,
+    size_t *oldlenp, const void *newp, size_t newlen);
+#define ABORT_ON_TAG_LOSS_SYSCTL "security.cheri.abort_on_memcpy_tag_loss"
+#define ABORT_ON_TAG_LOSS_ENVVAR "CHERI_ABORT_ON_TAG_STRIPPING_COPY"
+
+static void
+handle_untagged_copy(const void *__capability taggedcap, vaddr_t src_addr,
+    vaddr_t dst_addr, size_t offset)
+{
+	char errmsg_buffer[1024];
+	/* XXXAR: These functions do not exist yet... */
+	snprintf(errmsg_buffer, sizeof(errmsg_buffer),
+	    "%s: Attempting to copy a tagged capability (%#p) from 0x%jx to "
+	    "underaligned destination 0x%jx. Use memmove_nocap()/memcpy_nocap()"
+	    " if you intended to strip tags.\n", getprogname(),
+#ifdef __CHERI_PURE_CAPABILITY__
+	    taggedcap,
+#else
+	    /* Can't use capabilities in fprintf in hybrid mode */
+	    (void *)(uintptr_t)(__cheri_addr vaddr_t)(taggedcap),
+#endif
+	    (uintmax_t)(src_addr + offset), (uintmax_t)(dst_addr + offset));
+	write_to_stderr(errmsg_buffer);
+	/* TODO: allow overriding the behaviour with a function pointer? */
+	static uint32_t abort_on_tag_loss = -1;
+	if (abort_on_tag_loss == -1) {
+		const char *from_env = getenv(ABORT_ON_TAG_LOSS_ENVVAR);
+		if (from_env != NULL) {
+			/* Enabled unless empty or starts with zero. */
+			if (*from_env == '\0' || *from_env == '0')
+				abort_on_tag_loss = 0;
+			else
+				abort_on_tag_loss = 1;
+		}
+	}
+	if (abort_on_tag_loss == -1) {
+		/* If the env var is not set fall back to the global sysctl */
+		size_t olen = sizeof(abort_on_tag_loss);
+		if (__sysctlbyname(ABORT_ON_TAG_LOSS_SYSCTL,
+			strlen(ABORT_ON_TAG_LOSS_SYSCTL), &abort_on_tag_loss,
+			&olen, NULL, 0) == -1) {
+			write_to_stderr("ERROR: could not determine whether "
+			    "tag stripping memcpy should abort. Assuming it "
+			    "shouldn't.\n");
+			abort_on_tag_loss = 0;
+		}
+	}
```

Comment:
The global sysctl makes no sense, on CHERI-MIPS I always see a few in normal use so that would make processes crash. If you insist on having it have it as an environment variable only, otherwise you're giving people an option to shoot themselves in the foot (and they _will_ as it's confusingly in the "security" namespace).

Reply:
Those are bugs that we haven't bothered fixing (since it no longer abort()s). Why would copying unitialized data from the stack be useful? I get it that some people believe it's a source of entropy but please just use a tag-stripping memcpy().

Reply:
```
struct S {
    int flag;
    char buf[BUF_LEN];
};

void foo(char *str) {
    struct S s;
    s.flag = 1;
    strcpy(s.buf, str);
    bar(&s);
}

void bar(struct S *s) {
    char buf[BUF_LEN];
    memcpy(buf, s->buf);
    baz(buf);
}
```

Bit contrived but I hope you can see how there are legitimate coding patterns that can end up copying uninitialised stack data.

Reply:
Or, worse, if unions are involved you can have a _lot_ of padding that's uninitialised garbage.

Reply:
The union case I agree could be an issue (although C programmers very rarely pass those by value / copy them). Do you know what MSAN does there? Maybe opt-out of instrumenting those copies?

Reply:
I don't know, nor do I particularly care, because I believe it is fundamentally not memcpy's problem to solve.